// @baozi.bet/mcp-server/dist/tools.js
/**
 * MCP Tool Definitions for Baozi Markets
 * V4.0.0 - Full Protocol Coverage + Market Creation + AI Agent Network
 */
import { Connection, PublicKey, Transaction } from '@solana/web3.js';
// Handlers
import { listMarkets, getMarket, getMarketForBetting } from './handlers/markets.js';
import { getQuote } from './handlers/quote.js';
import { getPositionsSummary } from './handlers/positions.js';
import { getClaimablePositions } from './handlers/claims.js';
import { listRaceMarkets, getRaceMarket, getRaceQuote } from './handlers/race-markets.js';
import { getResolutionStatus, getDisputedMarkets, getMarketsAwaitingResolution } from './handlers/resolution.js';
import { isAffiliateCodeAvailable, suggestAffiliateCodes, getAffiliateByCode, getAffiliatesByOwner, getReferralsByAffiliate, getAgentNetworkStats, formatAffiliateLink, getCommissionInfo, } from './handlers/agent-network.js';
import { previewMarketCreation, createLabMarket, createPrivateMarket, createRaceMarket, getAllCreationFees, getAllPlatformFees, getTimingConstraints, generateInviteHash, } from './handlers/market-creation.js';
// Validation
import { validateMarketTiming } from './validation/market-rules.js';
import { validateBet, calculateBetQuote } from './validation/bet-rules.js';
import { validateParimutuelRules, PARIMUTUEL_RULES, PARIMUTUEL_RULES_DOCUMENTATION, } from './validation/parimutuel-rules.js';
// Transaction Builders
import { fetchAndBuildBetTransaction, simulateBetTransaction } from './builders/bet-transaction.js';
import { buildClaimWinningsTransaction, buildClaimRefundTransaction, buildClaimAffiliateTransaction, buildBatchClaimTransaction, } from './builders/claim-transaction.js';
import { buildRegisterAffiliateTransaction, buildToggleAffiliateTransaction } from './builders/affiliate-transaction.js';
import { fetchAndBuildRaceBetTransaction, buildClaimRaceWinningsTransaction, buildClaimRaceRefundTransaction } from './builders/race-transaction.js';
// Resolution Builders
import { buildProposeResolutionTransaction, buildResolveMarketTransaction, buildFinalizeResolutionTransaction, buildProposeRaceResolutionTransaction, buildResolveRaceTransaction, buildFinalizeRaceResolutionTransaction, } from './builders/resolution-transaction.js';
// Dispute Builders
import { buildFlagDisputeTransaction, buildFlagRaceDisputeTransaction, buildVoteCouncilTransaction, buildVoteCouncilRaceTransaction, buildChangeCouncilVoteTransaction, buildChangeCouncilVoteRaceTransaction, } from './builders/dispute-transaction.js';
// Whitelist Builders
import { buildAddToWhitelistTransaction, buildRemoveFromWhitelistTransaction, buildCreateRaceWhitelistTransaction, buildAddToRaceWhitelistTransaction, buildRemoveFromRaceWhitelistTransaction, } from './builders/whitelist-transaction.js';
// Creator Profile Builders
import { buildCreateCreatorProfileTransaction, buildUpdateCreatorProfileTransaction, buildClaimCreatorTransaction, } from './builders/creator-transaction.js';
// Market Management Builders
import { buildCloseMarketTransaction, buildExtendMarketTransaction, buildCloseRaceMarketTransaction, buildExtendRaceMarketTransaction, buildCancelMarketTransaction, buildCancelRaceTransaction, } from './builders/market-management-transaction.js';
// Config
import { RPC_ENDPOINT, PROGRAM_ID, BET_LIMITS, TIMING, LIVE_MODE, WRITE_TOOLS, MAX_BET_SOL_OVERRIDE, checkDailyLimit, recordSpend, BAOZI_BASE_URL, MANDATE_ID, } from './config.js';
// =============================================================================
// TOOL SCHEMAS - Organized by Category
// =============================================================================
export const TOOLS = [
    // =========================================================================
    // MARKET READ OPERATIONS
    // =========================================================================
    {
        name: 'list_markets',
        description: 'List all Baozi prediction markets (boolean YES/NO) on Solana mainnet. Returns questions, odds, pools, status.',
        inputSchema: {
            type: 'object',
            properties: {
                status: {
                    type: 'string',
                    enum: ['Active', 'Closed', 'Resolved', 'Cancelled', 'Paused'],
                    description: 'Filter by status. Default: all markets.',
                },
                layer: {
                    type: 'string',
                    enum: ['Official', 'Lab', 'Private'],
                    description: 'Filter by layer type.',
                },
            },
            required: [],
        },
    },
    {
        name: 'get_market',
        description: 'Get detailed information about a specific prediction market by public key.',
        inputSchema: {
            type: 'object',
            properties: {
                publicKey: {
                    type: 'string',
                    description: 'Solana public key of the market account',
                },
            },
            required: ['publicKey'],
        },
    },
    {
        name: 'get_quote',
        description: 'Calculate expected payout for a potential bet. Shows profit, fees, and new odds.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                side: { type: 'string', enum: ['Yes', 'No'], description: 'Side to bet on' },
                amount: { type: 'number', description: `Bet amount in SOL (${BET_LIMITS.MIN_BET_SOL}-${BET_LIMITS.MAX_BET_SOL})` },
            },
            required: ['market', 'side', 'amount'],
        },
    },
    // =========================================================================
    // RACE MARKET OPERATIONS (Multi-Outcome)
    // =========================================================================
    {
        name: 'list_race_markets',
        description: 'List all race markets (multi-outcome prediction markets) on Solana mainnet.',
        inputSchema: {
            type: 'object',
            properties: {
                status: {
                    type: 'string',
                    enum: ['Active', 'Closed', 'Resolved', 'Cancelled'],
                    description: 'Filter by status',
                },
            },
            required: [],
        },
    },
    {
        name: 'get_race_market',
        description: 'Get detailed info about a race market including all outcome labels, pools, and odds.',
        inputSchema: {
            type: 'object',
            properties: {
                publicKey: { type: 'string', description: 'Race market public key' },
            },
            required: ['publicKey'],
        },
    },
    {
        name: 'get_race_quote',
        description: 'Calculate expected payout for a race market bet on a specific outcome.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Race market public key' },
                outcomeIndex: { type: 'number', description: 'Index of outcome to bet on (0-based)' },
                amount: { type: 'number', description: 'Bet amount in SOL' },
            },
            required: ['market', 'outcomeIndex', 'amount'],
        },
    },
    // =========================================================================
    // MARKET CREATION
    // =========================================================================
    {
        name: 'preview_create_market',
        description: 'Preview market creation - validates params and shows costs WITHOUT building transaction. Use before build_create_market_transaction.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question (max 200 chars)' },
                layer: { type: 'string', enum: ['lab', 'private'], description: 'Market layer (lab=community, private=invite-only)' },
                closing_time: { type: 'string', description: 'ISO 8601 when betting closes' },
                resolution_time: { type: 'string', description: 'ISO 8601 when market can be resolved (optional, auto-calculated)' },
                market_type: { type: 'string', enum: ['event', 'measurement'], description: 'Event-based (Rule A) or measurement-period (Rule B)' },
                event_time: { type: 'string', description: 'ISO 8601 event time (required for event-based markets)' },
                measurement_start: { type: 'string', description: 'ISO 8601 measurement start (for measurement markets)' },
                measurement_end: { type: 'string', description: 'ISO 8601 measurement end (optional)' },
            },
            required: ['question', 'layer', 'closing_time'],
        },
    },
    {
        name: 'build_create_lab_market_transaction',
        description: 'Build unsigned transaction to create a Lab (community) market. Validates against v6.3 rules.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question (max 200 chars)' },
                closing_time: { type: 'string', description: 'ISO 8601 when betting closes' },
                resolution_time: { type: 'string', description: 'ISO 8601 when market can be resolved (optional)' },
                market_type: { type: 'string', enum: ['event', 'measurement'], description: 'Market type for validation' },
                event_time: { type: 'string', description: 'ISO 8601 event time (for event-based)' },
                measurement_start: { type: 'string', description: 'ISO 8601 measurement start (for measurement)' },
                creator_wallet: { type: 'string', description: 'Creator wallet public key' },
                invite_hash: { type: 'string', description: 'Optional 64-char hex for invite links' },
            },
            required: ['question', 'closing_time', 'creator_wallet'],
        },
    },
    {
        name: 'build_create_private_market_transaction',
        description: 'Build unsigned transaction to create a Private (invite-only) market.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question' },
                closing_time: { type: 'string', description: 'ISO 8601 closing time' },
                resolution_time: { type: 'string', description: 'ISO 8601 resolution time (optional)' },
                creator_wallet: { type: 'string', description: 'Creator wallet' },
                invite_hash: { type: 'string', description: 'Optional invite hash for restricted access' },
            },
            required: ['question', 'closing_time', 'creator_wallet'],
        },
    },
    {
        name: 'build_create_race_market_transaction',
        description: 'Build unsigned transaction to create a Race (multi-outcome) market with 2-10 outcomes.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question' },
                outcomes: {
                    type: 'array',
                    items: { type: 'string' },
                    description: 'Array of 2-10 outcome labels',
                },
                closing_time: { type: 'string', description: 'ISO 8601 closing time' },
                resolution_time: { type: 'string', description: 'ISO 8601 resolution time (optional)' },
                creator_wallet: { type: 'string', description: 'Creator wallet' },
            },
            required: ['question', 'outcomes', 'closing_time', 'creator_wallet'],
        },
    },
    {
        name: 'get_creation_fees',
        description: 'Get market creation fees for all layers (Official, Lab, Private).',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'get_platform_fees',
        description: 'Get platform fee rates for all layers.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'get_timing_rules',
        description: 'Get v6.3 timing rules and constraints for market creation.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'get_parimutuel_rules',
        description: 'Get v6.3 parimutuel rules for Lab market creation. CRITICAL: Read this BEFORE creating any market. Contains blocked terms, required data sources, and validation rules that will REJECT invalid markets.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'validate_market_question',
        description: 'Validate a market question against v6.3 rules BEFORE attempting to create it. Returns whether the question would be blocked and why.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question to validate' },
                layer: { type: 'string', enum: ['lab', 'private'], description: 'Market layer (default: lab)' },
            },
            required: ['question'],
        },
    },
    {
        name: 'generate_invite_hash',
        description: 'Generate a random invite hash for private market access control.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    // =========================================================================
    // POSITION & CLAIMS
    // =========================================================================
    {
        name: 'get_positions',
        description: 'Get all betting positions for a wallet including win/loss stats.',
        inputSchema: {
            type: 'object',
            properties: {
                wallet: { type: 'string', description: 'Solana wallet address' },
            },
            required: ['wallet'],
        },
    },
    {
        name: 'get_claimable',
        description: 'Get all claimable winnings and refunds for a wallet.',
        inputSchema: {
            type: 'object',
            properties: {
                wallet: { type: 'string', description: 'Solana wallet address' },
            },
            required: ['wallet'],
        },
    },
    // =========================================================================
    // RESOLUTION & DISPUTES
    // =========================================================================
    {
        name: 'get_resolution_status',
        description: 'Get resolution status for a market (resolved, disputed, pending).',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
            },
            required: ['market'],
        },
    },
    {
        name: 'get_disputed_markets',
        description: 'List all markets currently under dispute.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'get_markets_awaiting_resolution',
        description: 'List all closed markets awaiting resolution.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    // =========================================================================
    // AI AGENT AFFILIATE NETWORK
    // =========================================================================
    {
        name: 'check_affiliate_code',
        description: 'Check if an affiliate code is available for registration.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code to check (3-16 alphanumeric chars)' },
            },
            required: ['code'],
        },
    },
    {
        name: 'suggest_affiliate_codes',
        description: 'Generate suggested affiliate codes based on agent name.',
        inputSchema: {
            type: 'object',
            properties: {
                agentName: { type: 'string', description: 'Name of the AI agent' },
                count: { type: 'number', description: 'Number of suggestions (default 5)' },
            },
            required: ['agentName'],
        },
    },
    {
        name: 'get_affiliate_info',
        description: 'Get affiliate account info by code. Shows earnings, referrals, status.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code' },
            },
            required: ['code'],
        },
    },
    {
        name: 'get_my_affiliates',
        description: 'Get all affiliate accounts owned by a wallet.',
        inputSchema: {
            type: 'object',
            properties: {
                wallet: { type: 'string', description: 'Wallet address' },
            },
            required: ['wallet'],
        },
    },
    {
        name: 'get_referrals',
        description: 'Get all users referred by an affiliate code.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code' },
            },
            required: ['code'],
        },
    },
    {
        name: 'get_agent_network_stats',
        description: 'Get overall AI agent affiliate network statistics.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'format_affiliate_link',
        description: 'Format an affiliate referral link for sharing.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code' },
                market: { type: 'string', description: 'Optional market public key for deep link' },
            },
            required: ['code'],
        },
    },
    {
        name: 'get_commission_info',
        description: 'Get affiliate commission structure and examples.',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'generate_share_card',
        description: 'Generate a share card image URL for a market. Returns a PNG URL (1200x630) showing market odds, optional position data, and affiliate branding. Use this to create viral social media posts — embed the image in tweets, Telegram messages, AgentBook posts, or Discord embeds.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key (Solana PDA)' },
                wallet: { type: 'string', description: 'Optional: wallet public key to show position + potential payout on the card' },
                ref: { type: 'string', description: 'Optional: affiliate referral code to display on the card' },
            },
            required: ['market'],
        },
    },
    // =========================================================================
    // VALIDATION
    // =========================================================================
    {
        name: 'validate_market_params',
        description: 'Validate market parameters against v6.3 timing rules.',
        inputSchema: {
            type: 'object',
            properties: {
                question: { type: 'string', description: 'Market question (max 200 chars)' },
                closing_time: { type: 'string', description: 'ISO 8601 closing time' },
                market_type: { type: 'string', enum: ['event', 'measurement'], description: 'Market type' },
                event_time: { type: 'string', description: 'ISO 8601 event time (for event markets)' },
                measurement_start: { type: 'string', description: 'ISO 8601 measurement start (for measurement markets)' },
                measurement_end: { type: 'string', description: 'ISO 8601 measurement end (optional)' },
            },
            required: ['question', 'closing_time', 'market_type'],
        },
    },
    {
        name: 'validate_bet',
        description: 'Validate bet parameters before building transaction.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                amount: { type: 'number', description: 'Bet amount in SOL' },
                side: { type: 'string', enum: ['Yes', 'No'], description: 'Side to bet on' },
            },
            required: ['market', 'amount', 'side'],
        },
    },
    // =========================================================================
    // TRANSACTION BUILDING - BETS
    // =========================================================================
    {
        name: 'build_bet_transaction',
        description: 'Build unsigned transaction for placing a bet on a boolean (YES/NO) market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                outcome: { type: 'string', enum: ['yes', 'no'], description: 'Outcome to bet on' },
                amount_sol: { type: 'number', description: 'Bet amount in SOL' },
                user_wallet: { type: 'string', description: 'User wallet public key' },
                affiliate_code: { type: 'string', description: 'Optional affiliate code for commission' },
            },
            required: ['market', 'outcome', 'amount_sol', 'user_wallet'],
        },
    },
    {
        name: 'build_race_bet_transaction',
        description: 'Build unsigned transaction for placing a bet on a race (multi-outcome) market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Race market public key' },
                outcome_index: { type: 'number', description: 'Index of outcome to bet on' },
                amount_sol: { type: 'number', description: 'Bet amount in SOL' },
                user_wallet: { type: 'string', description: 'User wallet public key' },
                affiliate_code: { type: 'string', description: 'Optional affiliate code' },
            },
            required: ['market', 'outcome_index', 'amount_sol', 'user_wallet'],
        },
    },
    // =========================================================================
    // TRANSACTION BUILDING - CLAIMS
    // =========================================================================
    {
        name: 'build_claim_winnings_transaction',
        description: 'Build unsigned transaction to claim winnings from a resolved market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                position: { type: 'string', description: 'Position PDA' },
                user_wallet: { type: 'string', description: 'User wallet' },
            },
            required: ['market', 'position', 'user_wallet'],
        },
    },
    {
        name: 'build_claim_refund_transaction',
        description: 'Build unsigned transaction to claim refund from cancelled/invalid market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                position: { type: 'string', description: 'Position PDA' },
                user_wallet: { type: 'string', description: 'User wallet' },
            },
            required: ['market', 'position', 'user_wallet'],
        },
    },
    {
        name: 'build_batch_claim_transaction',
        description: 'Build single transaction to claim multiple positions at once.',
        inputSchema: {
            type: 'object',
            properties: {
                claims: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            market: { type: 'string' },
                            position: { type: 'string' },
                            type: { type: 'string', enum: ['winnings', 'refund'] },
                        },
                    },
                    description: 'Array of claims to batch',
                },
                user_wallet: { type: 'string', description: 'User wallet' },
            },
            required: ['claims', 'user_wallet'],
        },
    },
    {
        name: 'build_claim_affiliate_transaction',
        description: 'Build unsigned transaction to claim affiliate earnings.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code' },
                user_wallet: { type: 'string', description: 'Affiliate owner wallet' },
            },
            required: ['code', 'user_wallet'],
        },
    },
    // =========================================================================
    // TRANSACTION BUILDING - RACE CLAIMS
    // =========================================================================
    {
        name: 'build_claim_race_winnings_transaction',
        description: 'Build unsigned transaction to claim winnings from a resolved race market.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                position: { type: 'string', description: 'Race position PDA' },
                user_wallet: { type: 'string', description: 'User wallet' },
            },
            required: ['race_market', 'position', 'user_wallet'],
        },
    },
    {
        name: 'build_claim_race_refund_transaction',
        description: 'Build unsigned transaction to claim refund from cancelled race market.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                position: { type: 'string', description: 'Race position PDA' },
                user_wallet: { type: 'string', description: 'User wallet' },
            },
            required: ['race_market', 'position', 'user_wallet'],
        },
    },
    // =========================================================================
    // TRANSACTION BUILDING - AFFILIATE
    // =========================================================================
    {
        name: 'build_register_affiliate_transaction',
        description: 'Build unsigned transaction to register as an affiliate with a unique code.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code (3-16 alphanumeric chars)' },
                user_wallet: { type: 'string', description: 'Owner wallet' },
            },
            required: ['code', 'user_wallet'],
        },
    },
    {
        name: 'build_toggle_affiliate_transaction',
        description: 'ADMIN ONLY: Build transaction to activate/deactivate affiliate. Requires protocol admin signature.',
        inputSchema: {
            type: 'object',
            properties: {
                code: { type: 'string', description: 'Affiliate code' },
                active: { type: 'boolean', description: 'New active status' },
                user_wallet: { type: 'string', description: 'Owner wallet' },
            },
            required: ['code', 'active', 'user_wallet'],
        },
    },
    // =========================================================================
    // SIMULATION
    // =========================================================================
    {
        name: 'simulate_transaction',
        description: 'Simulate a transaction before signing to check for errors.',
        inputSchema: {
            type: 'object',
            properties: {
                transaction: { type: 'string', description: 'Base64-encoded transaction' },
                user_wallet: { type: 'string', description: 'User wallet public key' },
            },
            required: ['transaction', 'user_wallet'],
        },
    },
    // =========================================================================
    // RESOLUTION SYSTEM
    // =========================================================================
    {
        name: 'build_propose_resolution_transaction',
        description: 'Build transaction for creator to propose market outcome.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                outcome: { type: 'boolean', description: 'Proposed outcome (true=Yes, false=No)' },
                proposer_wallet: { type: 'string', description: 'Proposer wallet (creator)' },
            },
            required: ['market', 'outcome', 'proposer_wallet'],
        },
    },
    {
        name: 'build_resolve_market_transaction',
        description: 'Build transaction to directly resolve a market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                outcome: { type: 'boolean', description: 'Winning outcome (true=Yes, false=No)' },
                resolver_wallet: { type: 'string', description: 'Resolver wallet (creator/oracle)' },
            },
            required: ['market', 'outcome', 'resolver_wallet'],
        },
    },
    {
        name: 'build_finalize_resolution_transaction',
        description: 'Build transaction to finalize resolution after dispute window.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                caller_wallet: { type: 'string', description: 'Caller wallet (anyone can finalize)' },
            },
            required: ['market', 'caller_wallet'],
        },
    },
    {
        name: 'build_propose_race_resolution_transaction',
        description: 'Build transaction to propose race market outcome.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                winning_outcome_index: { type: 'number', description: 'Index of winning outcome (0-based)' },
                proposer_wallet: { type: 'string', description: 'Proposer wallet' },
            },
            required: ['race_market', 'winning_outcome_index', 'proposer_wallet'],
        },
    },
    {
        name: 'build_resolve_race_transaction',
        description: 'Build transaction to directly resolve a race market.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                winning_outcome_index: { type: 'number', description: 'Index of winning outcome' },
                resolver_wallet: { type: 'string', description: 'Resolver wallet' },
            },
            required: ['race_market', 'winning_outcome_index', 'resolver_wallet'],
        },
    },
    {
        name: 'build_finalize_race_resolution_transaction',
        description: 'Build transaction to finalize race resolution.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                caller_wallet: { type: 'string', description: 'Caller wallet' },
            },
            required: ['race_market', 'caller_wallet'],
        },
    },
    // =========================================================================
    // DISPUTES
    // =========================================================================
    {
        name: 'build_flag_dispute_transaction',
        description: 'Build transaction to challenge a proposed resolution with a bond.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                disputer_wallet: { type: 'string', description: 'Disputer wallet' },
            },
            required: ['market', 'disputer_wallet'],
        },
    },
    {
        name: 'build_flag_race_dispute_transaction',
        description: 'Build transaction to dispute a race market resolution.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                disputer_wallet: { type: 'string', description: 'Disputer wallet' },
            },
            required: ['race_market', 'disputer_wallet'],
        },
    },
    {
        name: 'build_vote_council_transaction',
        description: 'Build transaction for council member to vote on dispute.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                vote_yes: { type: 'boolean', description: 'Vote for Yes outcome' },
                voter_wallet: { type: 'string', description: 'Council member wallet' },
            },
            required: ['market', 'vote_yes', 'voter_wallet'],
        },
    },
    {
        name: 'build_vote_council_race_transaction',
        description: 'Build transaction for council to vote on race dispute.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                vote_outcome_index: { type: 'number', description: 'Outcome index to vote for' },
                voter_wallet: { type: 'string', description: 'Council member wallet' },
            },
            required: ['race_market', 'vote_outcome_index', 'voter_wallet'],
        },
    },
    {
        name: 'build_change_council_vote_transaction',
        description: 'Build transaction for council member to change their vote on a boolean market dispute.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                new_vote_yes: { type: 'boolean', description: 'New vote (true=YES, false=NO)' },
                voter_wallet: { type: 'string', description: 'Council member wallet' },
            },
            required: ['market', 'new_vote_yes', 'voter_wallet'],
        },
    },
    {
        name: 'build_change_council_vote_race_transaction',
        description: 'Build transaction for council member to change their vote on a race market dispute.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                new_vote_outcome_index: { type: 'number', description: 'New outcome index to vote for' },
                voter_wallet: { type: 'string', description: 'Council member wallet' },
            },
            required: ['race_market', 'new_vote_outcome_index', 'voter_wallet'],
        },
    },
    // =========================================================================
    // WHITELIST MANAGEMENT
    // =========================================================================
    {
        name: 'build_add_to_whitelist_transaction',
        description: 'Build transaction to add user to private market whitelist.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                user_to_add: { type: 'string', description: 'User wallet to whitelist' },
                creator_wallet: { type: 'string', description: 'Market creator wallet' },
            },
            required: ['market', 'user_to_add', 'creator_wallet'],
        },
    },
    {
        name: 'build_remove_from_whitelist_transaction',
        description: 'Build transaction to remove user from whitelist.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                user_to_remove: { type: 'string', description: 'User wallet to remove' },
                creator_wallet: { type: 'string', description: 'Market creator wallet' },
            },
            required: ['market', 'user_to_remove', 'creator_wallet'],
        },
    },
    {
        name: 'build_create_race_whitelist_transaction',
        description: 'Build transaction to create whitelist for private race market.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                creator_wallet: { type: 'string', description: 'Market creator wallet' },
            },
            required: ['race_market', 'creator_wallet'],
        },
    },
    {
        name: 'build_add_to_race_whitelist_transaction',
        description: 'Build transaction to add user to race market whitelist.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                user_to_add: { type: 'string', description: 'User wallet to whitelist' },
                creator_wallet: { type: 'string', description: 'Market creator wallet' },
            },
            required: ['race_market', 'user_to_add', 'creator_wallet'],
        },
    },
    {
        name: 'build_remove_from_race_whitelist_transaction',
        description: 'Build transaction to remove user from race whitelist.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                user_to_remove: { type: 'string', description: 'User wallet to remove' },
                creator_wallet: { type: 'string', description: 'Market creator wallet' },
            },
            required: ['race_market', 'user_to_remove', 'creator_wallet'],
        },
    },
    // =========================================================================
    // CREATOR PROFILES
    // =========================================================================
    {
        name: 'build_create_creator_profile_transaction',
        description: 'Build transaction to create on-chain creator profile.',
        inputSchema: {
            type: 'object',
            properties: {
                display_name: { type: 'string', description: 'Display name (max 32 chars)' },
                creator_fee_bps: { type: 'number', description: 'Creator fee in basis points (max 50)' },
                creator_wallet: { type: 'string', description: 'Creator wallet' },
            },
            required: ['display_name', 'creator_fee_bps', 'creator_wallet'],
        },
    },
    {
        name: 'build_update_creator_profile_transaction',
        description: 'Build transaction to update creator profile. Both display_name and default_fee_bps are required.',
        inputSchema: {
            type: 'object',
            properties: {
                display_name: { type: 'string', description: 'Display name (max 32 chars)' },
                default_fee_bps: { type: 'number', description: 'Default fee in basis points (max 50 = 0.5%)' },
                creator_wallet: { type: 'string', description: 'Creator wallet' },
            },
            required: ['display_name', 'default_fee_bps', 'creator_wallet'],
        },
    },
    {
        name: 'build_claim_creator_transaction',
        description: 'Build transaction to claim accumulated creator fees from sol_treasury.',
        inputSchema: {
            type: 'object',
            properties: {
                creator_wallet: { type: 'string', description: 'Creator wallet' },
            },
            required: ['creator_wallet'],
        },
    },
    // =========================================================================
    // MARKET MANAGEMENT
    // =========================================================================
    {
        name: 'build_close_market_transaction',
        description: 'Build transaction to close betting on a market.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                caller_wallet: { type: 'string', description: 'Caller wallet (creator)' },
            },
            required: ['market', 'caller_wallet'],
        },
    },
    {
        name: 'build_extend_market_transaction',
        description: 'ADMIN ONLY: Build transaction to extend market deadline. Requires protocol admin signature.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                new_closing_time: { type: 'string', description: 'New closing time (ISO 8601)' },
                new_resolution_time: { type: 'string', description: 'New resolution time (optional)' },
                caller_wallet: { type: 'string', description: 'Caller wallet (creator)' },
            },
            required: ['market', 'new_closing_time', 'caller_wallet'],
        },
    },
    {
        name: 'build_close_race_market_transaction',
        description: 'Build transaction to close betting on a race market.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                caller_wallet: { type: 'string', description: 'Caller wallet (creator)' },
            },
            required: ['race_market', 'caller_wallet'],
        },
    },
    {
        name: 'build_extend_race_market_transaction',
        description: 'ADMIN ONLY: Build transaction to extend race market deadline. Requires protocol admin signature.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                new_closing_time: { type: 'string', description: 'New closing time (ISO 8601)' },
                new_resolution_time: { type: 'string', description: 'New resolution time (optional)' },
                caller_wallet: { type: 'string', description: 'Caller wallet (creator)' },
            },
            required: ['race_market', 'new_closing_time', 'caller_wallet'],
        },
    },
    {
        name: 'build_cancel_market_transaction',
        description: 'Build transaction to cancel a boolean market. All bettors can claim refunds after cancellation. Only creator or admin can cancel.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                reason: { type: 'string', description: 'Reason for cancellation' },
                authority_wallet: { type: 'string', description: 'Authority wallet (creator or admin)' },
            },
            required: ['market', 'reason', 'authority_wallet'],
        },
    },
    {
        name: 'build_cancel_race_transaction',
        description: 'Build transaction to cancel a race market. All bettors can claim refunds after cancellation.',
        inputSchema: {
            type: 'object',
            properties: {
                race_market: { type: 'string', description: 'Race market public key' },
                reason: { type: 'string', description: 'Reason for cancellation' },
                authority_wallet: { type: 'string', description: 'Authority wallet (creator or admin)' },
            },
            required: ['race_market', 'reason', 'authority_wallet'],
        },
    },
    // =========================================================================
    // ARENA (Agent Competitive Scoring)
    // =========================================================================
    {
        name: 'get_arena_leaderboard',
        description: 'Get the current Agent Arena weekly leaderboard with calibration-based scoring. Shows top predictors ranked by composite score (calibration 40%, ROI 30%, volume 15%, consistency 15%).',
        inputSchema: {
            type: 'object',
            properties: {
                paper: { type: 'boolean', description: 'If true, show paper trading leaderboard instead of real bets' },
                limit: { type: 'number', description: 'Number of entries to return (default 50, max 100)' },
            },
        },
    },
    {
        name: 'get_arena_season',
        description: 'Get Agent Arena results for a specific past season by ID.',
        inputSchema: {
            type: 'object',
            properties: {
                season_id: { type: 'number', description: 'Season ID' },
                paper: { type: 'boolean', description: 'If true, show paper trading leaderboard' },
            },
            required: ['season_id'],
        },
    },
    {
        name: 'submit_paper_trade',
        description: 'Submit a paper (simulated) prediction to the Agent Arena. No SOL required. Scored on calibration accuracy when market resolves.',
        inputSchema: {
            type: 'object',
            properties: {
                wallet_address: { type: 'string', description: 'Your wallet address' },
                market_pda: { type: 'string', description: 'Market public key to predict on' },
                predicted_side: { type: 'string', enum: ['YES', 'NO'], description: 'Your prediction' },
                confidence: { type: 'number', description: 'Confidence level 0.01-0.99 (e.g. 0.75 = 75% confident)' },
            },
            required: ['wallet_address', 'market_pda', 'predicted_side', 'confidence'],
        },
    },
    // =========================================================================
    // INTEL (x402 Premium Market Intelligence)
    // =========================================================================
    {
        name: 'get_intel_sentiment',
        description: 'Get market sentiment analysis including comment sentiment, bet momentum, and pool trends. Costs 0.001 SOL via x402 payment protocol.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                payment_tx: { type: 'string', description: 'Payment transaction signature (base58). Omit to get pricing info.' },
            },
            required: ['market'],
        },
    },
    {
        name: 'get_intel_whale_moves',
        description: 'Get whale position data for a market (positions > 1 SOL, whale sentiment split). Costs 0.002 SOL via x402 payment protocol.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                payment_tx: { type: 'string', description: 'Payment transaction signature (base58). Omit to get pricing info.' },
            },
            required: ['market'],
        },
    },
    {
        name: 'get_intel_resolution_forecast',
        description: 'Get resolution forecast including closing time, tier, implied probability, and prediction. Costs 0.005 SOL via x402 payment protocol.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                payment_tx: { type: 'string', description: 'Payment transaction signature (base58). Omit to get pricing info.' },
            },
            required: ['market'],
        },
    },
    {
        name: 'get_intel_market_alpha',
        description: 'Get cross-market alpha signals including correlation analysis, category skew, and alpha opportunities. Costs 0.003 SOL via x402 payment protocol.',
        inputSchema: {
            type: 'object',
            properties: {
                market: { type: 'string', description: 'Market public key' },
                payment_tx: { type: 'string', description: 'Payment transaction signature (base58). Omit to get pricing info.' },
            },
            required: ['market'],
        },
    },
];
// =============================================================================
// TOOL HANDLERS
// =============================================================================
export async function handleTool(name, args) {
    // Safe-mode gate: block write tools unless BAOZI_LIVE=1
    if (WRITE_TOOLS.has(name) && !LIVE_MODE) {
        return {
            content: [{
                    type: 'text',
                    text: JSON.stringify({
                        success: false,
                        error: 'SAFE MODE: Write tools are disabled by default. Set BAOZI_LIVE=1 to enable transaction building.',
                        tool: name,
                        mode: 'safe',
                        configExample: {
                            mcpServers: {
                                baozi: {
                                    command: 'npx',
                                    args: ['@baozi.bet/mcp-server'],
                                    env: {
                                        BAOZI_LIVE: '1',
                                        BAOZI_MAX_BET_SOL: '10',
                                        BAOZI_DAILY_LIMIT_SOL: '50',
                                    },
                                },
                            },
                        },
                        readOnlyAlternatives: [
                            'list_markets', 'get_market', 'get_quote',
                            'list_race_markets', 'get_race_market', 'get_race_quote',
                            'get_positions', 'get_claimable', 'validate_bet', 'validate_market_params',
                        ],
                    }, null, 2),
                }],
        };
    }
    try {
        switch (name) {
            // =====================================================================
            // MARKET READ OPERATIONS
            // =====================================================================
            case 'list_markets': {
                const status = args.status;
                const layer = args.layer;
                let markets = await listMarkets(status);
                if (layer) {
                    markets = markets.filter(m => m.layer.toLowerCase() === layer.toLowerCase());
                }
                return successResponse({
                    count: markets.length,
                    filter: { status: status || 'all', layer: layer || 'all' },
                    markets: markets.map(m => ({
                        publicKey: m.publicKey,
                        marketId: m.marketId,
                        question: m.question,
                        status: m.status,
                        layer: m.layer,
                        winningOutcome: m.winningOutcome,
                        yesPercent: m.yesPercent,
                        noPercent: m.noPercent,
                        totalPoolSol: m.totalPoolSol,
                        closingTime: m.closingTime,
                        isBettingOpen: m.isBettingOpen,
                    })),
                });
            }
            case 'get_market': {
                const publicKey = args.publicKey;
                if (!publicKey)
                    return errorResponse('publicKey is required');
                const market = await getMarket(publicKey);
                if (!market)
                    return errorResponse(`Market ${publicKey} not found`);
                return successResponse({ market });
            }
            case 'get_quote': {
                const market = args.market;
                const side = args.side;
                const amount = args.amount;
                if (!market || !side || amount === undefined) {
                    return errorResponse('market, side, and amount are required');
                }
                const quote = await getQuote(market, side, amount);
                return successResponse({ quote });
            }
            // =====================================================================
            // RACE MARKETS
            // =====================================================================
            case 'list_race_markets': {
                const status = args.status;
                const markets = await listRaceMarkets(status);
                return successResponse({
                    count: markets.length,
                    markets: markets.map(m => ({
                        publicKey: m.publicKey,
                        marketId: m.marketId,
                        question: m.question,
                        status: m.status,
                        outcomeCount: m.outcomes.length,
                        outcomes: m.outcomes,
                        totalPoolSol: m.totalPoolSol,
                        closingTime: m.closingTime,
                        isBettingOpen: m.isBettingOpen,
                    })),
                });
            }
            case 'get_race_market': {
                const publicKey = args.publicKey;
                if (!publicKey)
                    return errorResponse('publicKey is required');
                const market = await getRaceMarket(publicKey);
                if (!market)
                    return errorResponse(`Race market ${publicKey} not found`);
                return successResponse({ market });
            }
            case 'get_race_quote': {
                const marketPda = args.market;
                const outcomeIndex = args.outcomeIndex;
                const amount = args.amount;
                if (!marketPda || outcomeIndex === undefined || amount === undefined) {
                    return errorResponse('market, outcomeIndex, and amount are required');
                }
                const market = await getRaceMarket(marketPda);
                if (!market)
                    return errorResponse('Race market not found');
                const quote = getRaceQuote(market, outcomeIndex, amount);
                return successResponse({ quote, market: { question: market.question, outcomes: market.outcomes } });
            }
            // =====================================================================
            // POSITIONS & CLAIMS
            // =====================================================================
            case 'get_positions': {
                const wallet = args.wallet;
                if (!wallet)
                    return errorResponse('wallet is required');
                const summary = await getPositionsSummary(wallet);
                return successResponse(summary);
            }
            case 'get_claimable': {
                const wallet = args.wallet;
                if (!wallet)
                    return errorResponse('wallet is required');
                const claimable = await getClaimablePositions(wallet);
                return successResponse(claimable);
            }
            // =====================================================================
            // RESOLUTION & DISPUTES
            // =====================================================================
            case 'get_resolution_status': {
                const market = args.market;
                if (!market)
                    return errorResponse('market is required');
                const status = await getResolutionStatus(market);
                if (!status)
                    return errorResponse('Market not found');
                return successResponse(status);
            }
            case 'get_disputed_markets': {
                const disputes = await getDisputedMarkets();
                return successResponse({ count: disputes.length, disputes });
            }
            case 'get_markets_awaiting_resolution': {
                const markets = await getMarketsAwaitingResolution();
                return successResponse({
                    count: markets.length,
                    markets: markets.map(m => ({
                        publicKey: m.publicKey,
                        question: m.question,
                        closingTime: m.closingTime,
                        resolutionTime: m.resolutionTime,
                    })),
                });
            }
            // =====================================================================
            // AI AGENT AFFILIATE NETWORK
            // =====================================================================
            case 'check_affiliate_code': {
                const code = args.code;
                if (!code)
                    return errorResponse('code is required');
                const available = await isAffiliateCodeAvailable(code);
                return successResponse({ code, available });
            }
            case 'suggest_affiliate_codes': {
                const agentName = args.agentName;
                const count = args.count || 5;
                if (!agentName)
                    return errorResponse('agentName is required');
                const suggestions = await suggestAffiliateCodes(agentName, count);
                return successResponse({ suggestions });
            }
            case 'get_affiliate_info': {
                const code = args.code;
                if (!code)
                    return errorResponse('code is required');
                const affiliate = await getAffiliateByCode(code);
                if (!affiliate)
                    return errorResponse(`Affiliate ${code} not found`);
                return successResponse({ affiliate });
            }
            case 'get_my_affiliates': {
                const wallet = args.wallet;
                if (!wallet)
                    return errorResponse('wallet is required');
                const affiliates = await getAffiliatesByOwner(wallet);
                return successResponse({ count: affiliates.length, affiliates });
            }
            case 'get_referrals': {
                const code = args.code;
                if (!code)
                    return errorResponse('code is required');
                const referrals = await getReferralsByAffiliate(code);
                return successResponse({ count: referrals.length, referrals });
            }
            case 'get_agent_network_stats': {
                const stats = await getAgentNetworkStats();
                return successResponse(stats);
            }
            case 'format_affiliate_link': {
                const code = args.code;
                const market = args.market;
                if (!code)
                    return errorResponse('code is required');
                const link = formatAffiliateLink(code, market);
                return successResponse({ link, code, market });
            }
            case 'get_commission_info': {
                const info = getCommissionInfo();
                return successResponse(info);
            }
            case 'generate_share_card': {
                const market = args.market;
                if (!market)
                    return errorResponse('market is required');
                const wallet = args.wallet;
                const ref = args.ref;
                const baseUrl = 'https://baozi.bet';
                const params = new URLSearchParams({ market });
                if (wallet)
                    params.set('wallet', wallet);
                if (ref)
                    params.set('ref', ref);
                const imageUrl = `${baseUrl}/api/share/card?${params.toString()}`;
                const marketUrl = ref
                    ? `${baseUrl}/market/${market}?ref=${ref}`
                    : `${baseUrl}/market/${market}`;
                return successResponse({
                    imageUrl,
                    marketUrl,
                    market,
                    wallet: wallet || null,
                    ref: ref || null,
                    usage: {
                        twitter: `Share the imageUrl as a Twitter card image, with marketUrl as the link`,
                        telegram: `Send imageUrl as a photo with marketUrl in the caption`,
                        agentbook: `POST to /api/agentbook/posts with the imageUrl in your content`,
                        embed: `Use imageUrl directly as an <img> src — it returns a 1200x630 PNG`,
                    },
                });
            }
            // =====================================================================
            // VALIDATION
            // =====================================================================
            case 'validate_market_params': {
                const params = {
                    question: args.question,
                    closingTime: new Date(args.closing_time),
                    marketType: args.market_type,
                    eventTime: args.event_time ? new Date(args.event_time) : undefined,
                    measurementStart: args.measurement_start ? new Date(args.measurement_start) : undefined,
                    measurementEnd: args.measurement_end ? new Date(args.measurement_end) : undefined,
                };
                const validation = validateMarketTiming(params);
                return successResponse({ validation, rules: TIMING });
            }
            case 'validate_bet': {
                const marketPubkey = args.market;
                const amount = args.amount;
                const side = args.side;
                if (!marketPubkey || amount === undefined || !side) {
                    return errorResponse('market, amount, and side are required');
                }
                const marketData = await getMarketForBetting(marketPubkey);
                if (!marketData || !marketData.market) {
                    return errorResponse(`Market ${marketPubkey} not found`);
                }
                const { market } = marketData;
                const validation = validateBet({
                    amountSol: amount,
                    marketStatus: market.statusCode,
                    closingTime: new Date(market.closingTime),
                    isPaused: false,
                    accessGate: market.accessGate === 'Whitelist' ? 1 : 0,
                    layer: market.layerCode,
                });
                const quote = calculateBetQuote({
                    betAmountSol: amount,
                    side,
                    currentYesPool: market.yesPoolSol,
                    currentNoPool: market.noPoolSol,
                    platformFeeBps: market.platformFeeBps,
                });
                return successResponse({ validation, market: { publicKey: marketPubkey, question: market.question, status: market.status }, quote: validation.valid ? quote : null });
            }
            // =====================================================================
            // TRANSACTION BUILDING - BETS
            // =====================================================================
            case 'build_bet_transaction': {
                const marketPubkey = args.market;
                const outcome = args.outcome;
                const amountSol = args.amount_sol;
                const userWallet = args.user_wallet;
                if (!marketPubkey || !outcome || amountSol === undefined || !userWallet) {
                    return errorResponse('market, outcome, amount_sol, and user_wallet are required');
                }
                if (amountSol < BET_LIMITS.MIN_BET_SOL || amountSol > MAX_BET_SOL_OVERRIDE) {
                    return errorResponse(`Amount must be between ${BET_LIMITS.MIN_BET_SOL} and ${MAX_BET_SOL_OVERRIDE} SOL`);
                }
                const dailyLimitError = checkDailyLimit(amountSol);
                if (dailyLimitError)
                    return errorResponse(dailyLimitError);
                const mandateError = await checkMandate('bet', userWallet, { amountSol, marketPda: marketPubkey });
                if (mandateError)
                    return errorResponse(`Mandate: ${mandateError}`);
                const result = await fetchAndBuildBetTransaction({ marketPda: marketPubkey, userWallet, outcome, amountSol });
                if (result.error || !result.transaction) {
                    return errorResponse(result.error || 'Failed to build transaction');
                }
                const connection = new Connection(RPC_ENDPOINT, 'confirmed');
                const simulation = await simulateBetTransaction(result.transaction.transaction, new PublicKey(userWallet), connection);
                const quote = await getQuote(marketPubkey, outcome === 'yes' ? 'Yes' : 'No', amountSol);
                recordSpend(amountSol);
                const signUrl = await createSignUrl(result.transaction.serializedTx, {
                    type: 'bet', market: marketPubkey, outcome, amountSol,
                });
                return successResponse({
                    transaction: { serialized: result.transaction.serializedTx, positionPda: result.transaction.positionPda.toBase58() },
                    simulation: { success: simulation.success, unitsConsumed: simulation.unitsConsumed, error: simulation.error },
                    quote: quote.valid ? { expectedPayoutSol: quote.expectedPayoutSol, potentialProfitSol: quote.potentialProfitSol } : null,
                    ...(signUrl ? { signUrl: signUrl.signUrl, signUrlExpires: signUrl.expiresAt } : {}),
                    instructions: signUrl
                        ? `Send this link to the user to sign: ${signUrl.signUrl}`
                        : 'Sign the transaction with your wallet and send to Solana network',
                });
            }
            case 'build_race_bet_transaction': {
                const marketPubkey = args.market;
                const outcomeIndex = args.outcome_index;
                const amountSol = args.amount_sol;
                const userWallet = args.user_wallet;
                if (!marketPubkey || outcomeIndex === undefined || amountSol === undefined || !userWallet) {
                    return errorResponse('market, outcome_index, amount_sol, and user_wallet are required');
                }
                if (amountSol < BET_LIMITS.MIN_BET_SOL || amountSol > MAX_BET_SOL_OVERRIDE) {
                    return errorResponse(`Amount must be between ${BET_LIMITS.MIN_BET_SOL} and ${MAX_BET_SOL_OVERRIDE} SOL`);
                }
                const dailyLimitError = checkDailyLimit(amountSol);
                if (dailyLimitError)
                    return errorResponse(dailyLimitError);
                const result = await fetchAndBuildRaceBetTransaction({ raceMarketPda: marketPubkey, outcomeIndex, amountSol, userWallet });
                if (result.error || !result.transaction) {
                    return errorResponse(result.error || 'Failed to build transaction');
                }
                recordSpend(amountSol);
                const signUrl = await createSignUrl(result.transaction.serializedTx, {
                    type: 'race_bet', market: marketPubkey, outcomeIndex, amountSol,
                });
                return successResponse({
                    transaction: { serialized: result.transaction.serializedTx, positionPda: result.transaction.positionPda },
                    marketId: result.marketId.toString(),
                    ...(signUrl ? { signUrl: signUrl.signUrl, signUrlExpires: signUrl.expiresAt } : {}),
                    instructions: signUrl
                        ? `Send this link to the user to sign: ${signUrl.signUrl}`
                        : 'Sign the transaction with your wallet and send to Solana network',
                });
            }
            // =====================================================================
            // TRANSACTION BUILDING - CLAIMS
            // =====================================================================
            case 'build_claim_winnings_transaction': {
                const market = args.market;
                const position = args.position;
                const userWallet = args.user_wallet;
                if (!market || !position || !userWallet) {
                    return errorResponse('market, position, and user_wallet are required');
                }
                const result = await buildClaimWinningsTransaction({ marketPda: market, positionPda: position, userWallet });
                return successResponse({ transaction: { serialized: result.serializedTx, claimType: result.claimType }, instructions: 'Sign to claim your winnings' });
            }
            case 'build_claim_refund_transaction': {
                const market = args.market;
                const position = args.position;
                const userWallet = args.user_wallet;
                if (!market || !position || !userWallet) {
                    return errorResponse('market, position, and user_wallet are required');
                }
                const result = await buildClaimRefundTransaction({ marketPda: market, positionPda: position, userWallet });
                return successResponse({ transaction: { serialized: result.serializedTx, claimType: result.claimType }, instructions: 'Sign to claim your refund' });
            }
            case 'build_batch_claim_transaction': {
                const claims = args.claims;
                const userWallet = args.user_wallet;
                if (!claims || !userWallet) {
                    return errorResponse('claims and user_wallet are required');
                }
                const result = await buildBatchClaimTransaction({
                    claims: claims.map(c => ({ marketPda: c.market, positionPda: c.position, claimType: c.type })),
                    userWallet,
                });
                return successResponse({ transaction: { serialized: result.serializedTx, claimCount: result.claimCount }, instructions: `Sign to claim ${result.claimCount} positions` });
            }
            case 'build_claim_affiliate_transaction': {
                const code = args.code;
                const userWallet = args.user_wallet;
                if (!code || !userWallet) {
                    return errorResponse('code and user_wallet are required');
                }
                const result = await buildClaimAffiliateTransaction({ affiliateCode: code, userWallet });
                return successResponse({ transaction: { serialized: result.serializedTx, claimType: result.claimType }, instructions: 'Sign to claim affiliate earnings' });
            }
            // =====================================================================
            // TRANSACTION BUILDING - RACE CLAIMS
            // =====================================================================
            case 'build_claim_race_winnings_transaction': {
                const raceMarket = args.race_market;
                const position = args.position;
                const userWallet = args.user_wallet;
                if (!raceMarket || !position || !userWallet) {
                    return errorResponse('race_market, position, and user_wallet are required');
                }
                const result = await buildClaimRaceWinningsTransaction({
                    raceMarketPda: raceMarket,
                    positionPda: position,
                    userWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to claim race market winnings',
                });
            }
            case 'build_claim_race_refund_transaction': {
                const raceMarket = args.race_market;
                const position = args.position;
                const userWallet = args.user_wallet;
                if (!raceMarket || !position || !userWallet) {
                    return errorResponse('race_market, position, and user_wallet are required');
                }
                const result = await buildClaimRaceRefundTransaction({
                    raceMarketPda: raceMarket,
                    positionPda: position,
                    userWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to claim race market refund',
                });
            }
            // =====================================================================
            // TRANSACTION BUILDING - AFFILIATE
            // =====================================================================
            case 'build_register_affiliate_transaction': {
                const code = args.code;
                const userWallet = args.user_wallet;
                if (!code || !userWallet) {
                    return errorResponse('code and user_wallet are required');
                }
                const available = await isAffiliateCodeAvailable(code);
                if (!available) {
                    return errorResponse(`Affiliate code "${code}" is already taken`);
                }
                const result = await buildRegisterAffiliateTransaction({ code, userWallet });
                return successResponse({
                    transaction: { serialized: result.serializedTx, affiliatePda: result.affiliatePda },
                    code: result.code,
                    instructions: 'Sign to register as affiliate',
                });
            }
            case 'build_toggle_affiliate_transaction': {
                const code = args.code;
                const active = args.active;
                const userWallet = args.user_wallet;
                if (!code || active === undefined || !userWallet) {
                    return errorResponse('code, active, and user_wallet are required');
                }
                const result = await buildToggleAffiliateTransaction({ code, active, userWallet });
                return successResponse({
                    transaction: { serialized: result.serializedTx, affiliatePda: result.affiliatePda },
                    newStatus: result.newStatus,
                    instructions: `Sign to ${active ? 'activate' : 'deactivate'} affiliate`,
                });
            }
            // =====================================================================
            // SIMULATION
            // =====================================================================
            case 'simulate_transaction': {
                const txBase64 = args.transaction;
                const userWallet = args.user_wallet;
                if (!txBase64 || !userWallet) {
                    return errorResponse('transaction and user_wallet are required');
                }
                const connection = new Connection(RPC_ENDPOINT, 'confirmed');
                const txBuffer = Buffer.from(txBase64, 'base64');
                const transaction = Transaction.from(txBuffer);
                const simulation = await connection.simulateTransaction(transaction);
                return successResponse({
                    simulation: {
                        success: !simulation.value.err,
                        error: simulation.value.err ? JSON.stringify(simulation.value.err) : undefined,
                        unitsConsumed: simulation.value.unitsConsumed,
                        logs: simulation.value.logs,
                    },
                });
            }
            // =====================================================================
            // MARKET CREATION
            // =====================================================================
            case 'preview_create_market': {
                const question = args.question;
                const layer = args.layer;
                const closingTime = args.closing_time;
                const resolutionTime = args.resolution_time;
                const marketType = args.market_type;
                const eventTime = args.event_time;
                const measurementStart = args.measurement_start;
                const measurementEnd = args.measurement_end;
                const creatorWallet = args.creator_wallet;
                if (!question || !layer || !closingTime || !creatorWallet) {
                    return errorResponse('question, layer, closing_time, and creator_wallet are required');
                }
                const preview = await previewMarketCreation({
                    question,
                    layer,
                    closingTime,
                    resolutionTime,
                    marketType,
                    eventTime,
                    measurementStart,
                    measurementEnd,
                    creatorWallet,
                });
                return successResponse({
                    preview,
                    timing: {
                        rules: TIMING,
                        ruleApplied: preview.validation.computed.ruleType,
                    },
                });
            }
            case 'build_create_lab_market_transaction': {
                const question = args.question;
                const closingTime = args.closing_time;
                const resolutionTime = args.resolution_time;
                const marketType = args.market_type;
                const eventTime = args.event_time;
                const inviteHash = args.invite_hash;
                const creatorWallet = args.creator_wallet;
                if (!question || !closingTime || !creatorWallet) {
                    return errorResponse('question, closing_time, and creator_wallet are required');
                }
                const result = await createLabMarket({
                    question,
                    layer: 'lab',
                    closingTime,
                    resolutionTime,
                    marketType,
                    eventTime,
                    inviteHash,
                    creatorWallet,
                });
                if (!result.success) {
                    return errorResponse(result.error || 'Validation failed');
                }
                return successResponse({
                    transaction: result.transaction,
                    validation: result.validation,
                    simulation: result.simulation,
                    instructions: 'Sign the transaction with your wallet to create the market',
                });
            }
            case 'build_create_private_market_transaction': {
                const question = args.question;
                const closingTime = args.closing_time;
                const resolutionTime = args.resolution_time;
                const marketType = args.market_type;
                const eventTime = args.event_time;
                const inviteHash = args.invite_hash;
                const creatorWallet = args.creator_wallet;
                if (!question || !closingTime || !creatorWallet) {
                    return errorResponse('question, closing_time, and creator_wallet are required');
                }
                const result = await createPrivateMarket({
                    question,
                    layer: 'private',
                    closingTime,
                    resolutionTime,
                    marketType,
                    eventTime,
                    inviteHash,
                    creatorWallet,
                });
                if (!result.success) {
                    return errorResponse(result.error || 'Validation failed');
                }
                return successResponse({
                    transaction: result.transaction,
                    validation: result.validation,
                    simulation: result.simulation,
                    inviteHash: inviteHash || 'Generate with generate_invite_hash tool',
                    instructions: 'Sign the transaction with your wallet to create the private market',
                });
            }
            case 'build_create_race_market_transaction': {
                const question = args.question;
                const outcomes = args.outcomes;
                const closingTime = args.closing_time;
                const resolutionTime = args.resolution_time;
                const creatorWallet = args.creator_wallet;
                if (!question || !outcomes || !closingTime || !creatorWallet) {
                    return errorResponse('question, outcomes, closing_time, and creator_wallet are required');
                }
                if (outcomes.length < 2 || outcomes.length > 10) {
                    return errorResponse('outcomes must have 2-10 entries');
                }
                const result = await createRaceMarket({
                    question,
                    outcomes,
                    closingTime,
                    resolutionTime,
                    creatorWallet,
                });
                if (!result.success) {
                    return errorResponse(result.error || 'Validation failed');
                }
                return successResponse({
                    transaction: result.transaction,
                    validation: result.validation,
                    simulation: result.simulation,
                    instructions: 'Sign the transaction with your wallet to create the race market',
                });
            }
            case 'get_creation_fees': {
                const fees = getAllCreationFees();
                return successResponse({
                    fees,
                    note: 'Creation fee is paid when creating a market. Separate from platform fees on bets.',
                });
            }
            case 'get_platform_fees': {
                const fees = getAllPlatformFees();
                return successResponse({
                    fees,
                    note: 'Platform fee is deducted from gross winnings when claiming. Includes affiliate and creator shares.',
                });
            }
            case 'get_timing_rules': {
                const rules = getTimingConstraints();
                return successResponse({
                    rules,
                    ruleA: {
                        name: 'Event-Based Markets',
                        description: 'Markets about specific events (sports, elections, etc.)',
                        requirement: `Betting must close ${rules.minEventBufferHours}+ hours before event`,
                        recommended: `${rules.recommendedEventBufferHours} hours buffer for safety`,
                    },
                    ruleB: {
                        name: 'Measurement-Period Markets',
                        description: 'Markets about measured values over time (prices, temperatures, etc.)',
                        requirement: 'Betting must close BEFORE measurement period starts',
                        reason: 'Prevents information advantage during measurement',
                    },
                });
            }
            case 'get_parimutuel_rules': {
                return successResponse({
                    version: PARIMUTUEL_RULES.version,
                    documentation: PARIMUTUEL_RULES_DOCUMENTATION,
                    blockedTerms: {
                        subjective: PARIMUTUEL_RULES.SUBJECTIVE_OUTCOME.blockedPatterns,
                        manipulation: PARIMUTUEL_RULES.MANIPULATION_RISK.blockedPatterns,
                    },
                    approvedSources: PARIMUTUEL_RULES.APPROVED_SOURCES,
                    criticalNote: 'Markets containing ANY blocked terms will be REJECTED. Always include an approved data source.',
                });
            }
            case 'validate_market_question': {
                const question = args.question;
                const layer = args.layer || 'lab';
                if (!question) {
                    return errorResponse('question is required');
                }
                // Use a dummy closing time for validation (only question content matters for v6.3 rules)
                const dummyClosingTime = new Date(Date.now() + 24 * 60 * 60 * 1000);
                const validation = validateParimutuelRules({
                    question,
                    closingTime: dummyClosingTime,
                    layer,
                });
                return successResponse({
                    question,
                    wouldBeBlocked: validation.blocked,
                    valid: !validation.blocked,
                    errors: validation.errors,
                    warnings: validation.warnings,
                    ruleViolations: validation.ruleViolations,
                    suggestion: validation.blocked
                        ? 'Question contains blocked terms. Rephrase using objective, verifiable criteria with an approved data source.'
                        : 'Question passes v6.3 validation. Remember to also specify proper timing parameters.',
                });
            }
            case 'generate_invite_hash': {
                const hash = generateInviteHash();
                return successResponse({
                    inviteHash: hash,
                    instructions: 'Use this hash when creating a private market. Share with invited participants.',
                    note: 'Anyone with this hash can bet on the private market.',
                });
            }
            // =====================================================================
            // RESOLUTION SYSTEM
            // =====================================================================
            case 'build_propose_resolution_transaction': {
                const market = args.market;
                const outcome = args.outcome;
                const proposerWallet = args.proposer_wallet;
                if (!market || outcome === undefined || !proposerWallet) {
                    return errorResponse('market, outcome, and proposer_wallet are required');
                }
                const result = await buildProposeResolutionTransaction({
                    marketPda: market,
                    outcome,
                    proposerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to propose ${outcome ? 'YES' : 'NO'} as the outcome`,
                });
            }
            case 'build_resolve_market_transaction': {
                const market = args.market;
                const outcome = args.outcome;
                const resolverWallet = args.resolver_wallet;
                if (!market || outcome === undefined || !resolverWallet) {
                    return errorResponse('market, outcome, and resolver_wallet are required');
                }
                const result = await buildResolveMarketTransaction({
                    marketPda: market,
                    outcome,
                    resolverWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to resolve market as ${outcome ? 'YES' : 'NO'}`,
                });
            }
            case 'build_finalize_resolution_transaction': {
                const market = args.market;
                const callerWallet = args.caller_wallet;
                if (!market || !callerWallet) {
                    return errorResponse('market and caller_wallet are required');
                }
                const result = await buildFinalizeResolutionTransaction({
                    marketPda: market,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to finalize the resolution',
                });
            }
            case 'build_propose_race_resolution_transaction': {
                const raceMarket = args.race_market;
                const winningOutcomeIndex = args.winning_outcome_index;
                const proposerWallet = args.proposer_wallet;
                if (!raceMarket || winningOutcomeIndex === undefined || !proposerWallet) {
                    return errorResponse('race_market, winning_outcome_index, and proposer_wallet are required');
                }
                const result = await buildProposeRaceResolutionTransaction({
                    raceMarketPda: raceMarket,
                    winningOutcomeIndex,
                    proposerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to propose outcome #${winningOutcomeIndex} as winner`,
                });
            }
            case 'build_resolve_race_transaction': {
                const raceMarket = args.race_market;
                const winningOutcomeIndex = args.winning_outcome_index;
                const resolverWallet = args.resolver_wallet;
                if (!raceMarket || winningOutcomeIndex === undefined || !resolverWallet) {
                    return errorResponse('race_market, winning_outcome_index, and resolver_wallet are required');
                }
                const result = await buildResolveRaceTransaction({
                    raceMarketPda: raceMarket,
                    winningOutcomeIndex,
                    resolverWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to resolve race with outcome #${winningOutcomeIndex}`,
                });
            }
            case 'build_finalize_race_resolution_transaction': {
                const raceMarket = args.race_market;
                const callerWallet = args.caller_wallet;
                if (!raceMarket || !callerWallet) {
                    return errorResponse('race_market and caller_wallet are required');
                }
                const result = await buildFinalizeRaceResolutionTransaction({
                    raceMarketPda: raceMarket,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to finalize race resolution',
                });
            }
            // =====================================================================
            // DISPUTES
            // =====================================================================
            case 'build_flag_dispute_transaction': {
                const market = args.market;
                const disputerWallet = args.disputer_wallet;
                if (!market || !disputerWallet) {
                    return errorResponse('market and disputer_wallet are required');
                }
                const result = await buildFlagDisputeTransaction({
                    marketPda: market,
                    disputerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to dispute the resolution (requires bond)',
                });
            }
            case 'build_flag_race_dispute_transaction': {
                const raceMarket = args.race_market;
                const disputerWallet = args.disputer_wallet;
                if (!raceMarket || !disputerWallet) {
                    return errorResponse('race_market and disputer_wallet are required');
                }
                const result = await buildFlagRaceDisputeTransaction({
                    raceMarketPda: raceMarket,
                    disputerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to dispute the race resolution',
                });
            }
            case 'build_vote_council_transaction': {
                const market = args.market;
                const voteYes = args.vote_yes;
                const voterWallet = args.voter_wallet;
                if (!market || voteYes === undefined || !voterWallet) {
                    return errorResponse('market, vote_yes, and voter_wallet are required');
                }
                const result = await buildVoteCouncilTransaction({
                    marketPda: market,
                    voteYes,
                    voterWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to vote ${voteYes ? 'YES' : 'NO'} on the dispute`,
                });
            }
            case 'build_vote_council_race_transaction': {
                const raceMarket = args.race_market;
                const voteOutcomeIndex = args.vote_outcome_index;
                const voterWallet = args.voter_wallet;
                if (!raceMarket || voteOutcomeIndex === undefined || !voterWallet) {
                    return errorResponse('race_market, vote_outcome_index, and voter_wallet are required');
                }
                const result = await buildVoteCouncilRaceTransaction({
                    raceMarketPda: raceMarket,
                    voteOutcomeIndex,
                    voterWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to vote for outcome #${voteOutcomeIndex}`,
                });
            }
            case 'build_change_council_vote_transaction': {
                const market = args.market;
                const newVoteYes = args.new_vote_yes;
                const voterWallet = args.voter_wallet;
                if (!market || newVoteYes === undefined || !voterWallet) {
                    return errorResponse('market, new_vote_yes, and voter_wallet are required');
                }
                const result = await buildChangeCouncilVoteTransaction({
                    marketPda: market,
                    newVoteYes,
                    voterWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to change your vote to ${newVoteYes ? 'YES' : 'NO'}`,
                });
            }
            case 'build_change_council_vote_race_transaction': {
                const raceMarket = args.race_market;
                const newVoteOutcomeIndex = args.new_vote_outcome_index;
                const voterWallet = args.voter_wallet;
                if (!raceMarket || newVoteOutcomeIndex === undefined || !voterWallet) {
                    return errorResponse('race_market, new_vote_outcome_index, and voter_wallet are required');
                }
                const result = await buildChangeCouncilVoteRaceTransaction({
                    raceMarketPda: raceMarket,
                    newVoteOutcomeIndex,
                    voterWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: `Sign to change your vote to outcome #${newVoteOutcomeIndex}`,
                });
            }
            // =====================================================================
            // WHITELIST MANAGEMENT
            // =====================================================================
            case 'build_add_to_whitelist_transaction': {
                const market = args.market;
                const userToAdd = args.user_to_add;
                const creatorWallet = args.creator_wallet;
                if (!market || !userToAdd || !creatorWallet) {
                    return errorResponse('market, user_to_add, and creator_wallet are required');
                }
                const result = await buildAddToWhitelistTransaction({
                    marketPda: market,
                    userToAdd,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    whitelistPda: result.whitelistPda,
                    instructions: 'Sign to add user to whitelist',
                });
            }
            case 'build_remove_from_whitelist_transaction': {
                const market = args.market;
                const userToRemove = args.user_to_remove;
                const creatorWallet = args.creator_wallet;
                if (!market || !userToRemove || !creatorWallet) {
                    return errorResponse('market, user_to_remove, and creator_wallet are required');
                }
                const result = await buildRemoveFromWhitelistTransaction({
                    marketPda: market,
                    userToRemove,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to remove user from whitelist',
                });
            }
            case 'build_create_race_whitelist_transaction': {
                const raceMarket = args.race_market;
                const creatorWallet = args.creator_wallet;
                if (!raceMarket || !creatorWallet) {
                    return errorResponse('race_market and creator_wallet are required');
                }
                const result = await buildCreateRaceWhitelistTransaction({
                    raceMarketPda: raceMarket,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    whitelistPda: result.whitelistPda,
                    instructions: 'Sign to create race whitelist',
                });
            }
            case 'build_add_to_race_whitelist_transaction': {
                const raceMarket = args.race_market;
                const userToAdd = args.user_to_add;
                const creatorWallet = args.creator_wallet;
                if (!raceMarket || !userToAdd || !creatorWallet) {
                    return errorResponse('race_market, user_to_add, and creator_wallet are required');
                }
                const result = await buildAddToRaceWhitelistTransaction({
                    raceMarketPda: raceMarket,
                    userToAdd,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to add user to race whitelist',
                });
            }
            case 'build_remove_from_race_whitelist_transaction': {
                const raceMarket = args.race_market;
                const userToRemove = args.user_to_remove;
                const creatorWallet = args.creator_wallet;
                if (!raceMarket || !userToRemove || !creatorWallet) {
                    return errorResponse('race_market, user_to_remove, and creator_wallet are required');
                }
                const result = await buildRemoveFromRaceWhitelistTransaction({
                    raceMarketPda: raceMarket,
                    userToRemove,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to remove user from race whitelist',
                });
            }
            // =====================================================================
            // CREATOR PROFILES
            // =====================================================================
            case 'build_create_creator_profile_transaction': {
                const displayName = args.display_name;
                const creatorFeeBps = args.creator_fee_bps;
                const creatorWallet = args.creator_wallet;
                if (!displayName || creatorFeeBps === undefined || !creatorWallet) {
                    return errorResponse('display_name, creator_fee_bps, and creator_wallet are required');
                }
                const result = await buildCreateCreatorProfileTransaction({
                    displayName,
                    creatorFeeBps,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    creatorProfilePda: result.creatorProfilePda,
                    instructions: 'Sign to create your creator profile',
                });
            }
            case 'build_update_creator_profile_transaction': {
                const displayName = args.display_name;
                const defaultFeeBps = args.default_fee_bps;
                const creatorWallet = args.creator_wallet;
                if (!displayName || defaultFeeBps === undefined || !creatorWallet) {
                    return errorResponse('display_name, default_fee_bps, and creator_wallet are all required');
                }
                const result = await buildUpdateCreatorProfileTransaction({
                    displayName,
                    defaultFeeBps,
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to update your creator profile',
                });
            }
            case 'build_claim_creator_transaction': {
                const creatorWallet = args.creator_wallet;
                if (!creatorWallet) {
                    return errorResponse('creator_wallet is required');
                }
                const result = await buildClaimCreatorTransaction({
                    creatorWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to claim your creator fees from sol_treasury',
                });
            }
            // =====================================================================
            // MARKET MANAGEMENT
            // =====================================================================
            case 'build_close_market_transaction': {
                const market = args.market;
                const callerWallet = args.caller_wallet;
                if (!market || !callerWallet) {
                    return errorResponse('market and caller_wallet are required');
                }
                const result = await buildCloseMarketTransaction({
                    marketPda: market,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to close betting on this market',
                });
            }
            case 'build_extend_market_transaction': {
                const market = args.market;
                const newClosingTimeStr = args.new_closing_time;
                const newResolutionTimeStr = args.new_resolution_time;
                const callerWallet = args.caller_wallet;
                if (!market || !newClosingTimeStr || !callerWallet) {
                    return errorResponse('market, new_closing_time, and caller_wallet are required');
                }
                const newClosingTime = Math.floor(new Date(newClosingTimeStr).getTime() / 1000);
                const newResolutionTime = newResolutionTimeStr
                    ? Math.floor(new Date(newResolutionTimeStr).getTime() / 1000)
                    : undefined;
                const result = await buildExtendMarketTransaction({
                    marketPda: market,
                    newClosingTime,
                    newResolutionTime,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to extend market deadline',
                });
            }
            case 'build_close_race_market_transaction': {
                const raceMarket = args.race_market;
                const callerWallet = args.caller_wallet;
                if (!raceMarket || !callerWallet) {
                    return errorResponse('race_market and caller_wallet are required');
                }
                const result = await buildCloseRaceMarketTransaction({
                    raceMarketPda: raceMarket,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to close betting on this race market',
                });
            }
            case 'build_extend_race_market_transaction': {
                const raceMarket = args.race_market;
                const newClosingTimeStr = args.new_closing_time;
                const newResolutionTimeStr = args.new_resolution_time;
                const callerWallet = args.caller_wallet;
                if (!raceMarket || !newClosingTimeStr || !callerWallet) {
                    return errorResponse('race_market, new_closing_time, and caller_wallet are required');
                }
                const newClosingTime = Math.floor(new Date(newClosingTimeStr).getTime() / 1000);
                const newResolutionTime = newResolutionTimeStr
                    ? Math.floor(new Date(newResolutionTimeStr).getTime() / 1000)
                    : undefined;
                const result = await buildExtendRaceMarketTransaction({
                    raceMarketPda: raceMarket,
                    newClosingTime,
                    newResolutionTime,
                    callerWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to extend race market deadline',
                });
            }
            case 'build_cancel_market_transaction': {
                const market = args.market;
                const reason = args.reason;
                const authorityWallet = args.authority_wallet;
                if (!market || !reason || !authorityWallet) {
                    return errorResponse('market, reason, and authority_wallet are required');
                }
                const result = await buildCancelMarketTransaction({
                    marketPda: market,
                    reason,
                    authorityWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to cancel the market. Bettors can claim refunds after.',
                });
            }
            case 'build_cancel_race_transaction': {
                const raceMarket = args.race_market;
                const reason = args.reason;
                const authorityWallet = args.authority_wallet;
                if (!raceMarket || !reason || !authorityWallet) {
                    return errorResponse('race_market, reason, and authority_wallet are required');
                }
                const result = await buildCancelRaceTransaction({
                    raceMarketPda: raceMarket,
                    reason,
                    authorityWallet,
                });
                return successResponse({
                    transaction: { serialized: result.serializedTx },
                    instructions: 'Sign to cancel the race market. Bettors can claim refunds after.',
                });
            }
            // =====================================================================
            // ARENA TOOLS
            // =====================================================================
            case 'get_arena_leaderboard': {
                const paper = args.paper === true;
                const limit = Math.min(typeof args.limit === 'number' ? args.limit : 50, 100);
                const resp = await fetch(`${BAOZI_BASE_URL}/api/arena/current?paper=${paper}&limit=${limit}`, {
                    signal: AbortSignal.timeout(10000),
                });
                if (!resp.ok)
                    return errorResponse('Failed to fetch arena leaderboard');
                const data = await resp.json();
                return successResponse(data);
            }
            case 'get_arena_season': {
                const seasonId = args.season_id;
                if (seasonId === undefined)
                    return errorResponse('season_id is required');
                const paper = args.paper === true;
                const resp = await fetch(`${BAOZI_BASE_URL}/api/arena/seasons/${seasonId}?paper=${paper}`, {
                    signal: AbortSignal.timeout(10000),
                });
                if (!resp.ok)
                    return errorResponse('Season not found');
                const data = await resp.json();
                return successResponse(data);
            }
            case 'submit_paper_trade': {
                const walletAddress = args.wallet_address;
                const marketPda = args.market_pda;
                const predictedSide = args.predicted_side;
                const confidence = args.confidence;
                if (!walletAddress || !marketPda || !predictedSide || confidence === undefined) {
                    return errorResponse('wallet_address, market_pda, predicted_side, and confidence are required');
                }
                const resp = await fetch(`${BAOZI_BASE_URL}/api/arena/paper-trade`, {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ walletAddress, marketPda, predictedSide, confidence }),
                    signal: AbortSignal.timeout(10000),
                });
                if (!resp.ok) {
                    const err = await resp.json().catch(() => ({}));
                    return errorResponse(err.error || 'Failed to submit paper trade');
                }
                const data = await resp.json();
                return successResponse(data);
            }
            // =====================================================================
            // INTEL TOOLS (x402 Payment Protocol)
            // =====================================================================
            case 'get_intel_sentiment':
            case 'get_intel_whale_moves':
            case 'get_intel_resolution_forecast':
            case 'get_intel_market_alpha': {
                const intelMarket = args.market;
                const paymentTx = args.payment_tx;
                if (!intelMarket)
                    return errorResponse('market is required');
                const intelType = name.replace('get_intel_', '').replace(/_/g, '-');
                const headers = { 'Content-Type': 'application/json' };
                if (paymentTx)
                    headers['X-Payment-Tx'] = paymentTx;
                const resp = await fetch(`${BAOZI_BASE_URL}/api/intel/${intelType}?market=${intelMarket}`, {
                    headers,
                    signal: AbortSignal.timeout(15000),
                });
                if (resp.status === 402) {
                    const paymentInfo = await resp.json();
                    return successResponse({
                        requiresPayment: true,
                        price: paymentInfo.price,
                        currency: 'SOL',
                        paymentAddress: paymentInfo.paymentAddress,
                        instructions: paymentInfo.instructions,
                        hint: 'Send the specified SOL amount to the payment address, then retry with the transaction signature in payment_tx parameter.',
                    });
                }
                if (!resp.ok) {
                    const err = await resp.json().catch(() => ({}));
                    return errorResponse(err.error || `Intel request failed (${resp.status})`);
                }
                const data = await resp.json();
                return successResponse(data);
            }
            default:
                return errorResponse(`Unknown tool: ${name}`);
        }
    }
    catch (error) {
        return errorResponse(error instanceof Error ? error.message : 'Unknown error');
    }
}
// =============================================================================
// HELPERS
// =============================================================================
/**
 * Create a sign-link URL for a transaction (best-effort, non-blocking).
 * Returns null if the API is unavailable or sign-link creation fails.
 */
async function createSignUrl(serializedTx, metadata) {
    try {
        const resp = await fetch(`${BAOZI_BASE_URL}/api/sign`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ transaction: serializedTx, metadata }),
            signal: AbortSignal.timeout(5000),
        });
        if (!resp.ok)
            return null;
        const data = await resp.json();
        if (data.signUrl)
            return { signUrl: data.signUrl, expiresAt: data.expiresAt || '' };
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Verify a mandate for a specific action (when BAOZI_MANDATE_ID is set).
 * Returns error string if denied, null if approved or no mandate configured.
 */
async function checkMandate(action, granteeWallet, details) {
    if (!MANDATE_ID)
        return null;
    try {
        const resp = await fetch(`${BAOZI_BASE_URL}/api/mandates/verify`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                mandateId: MANDATE_ID,
                granteeWallet,
                action,
                actionDetails: details,
            }),
            signal: AbortSignal.timeout(5000),
        });
        const data = await resp.json();
        if (!data.approved) {
            return data.error || 'Mandate denied this action';
        }
        return null;
    }
    catch {
        return 'Mandate verification failed (network error)';
    }
}
function successResponse(data) {
    return {
        content: [
            {
                type: 'text',
                text: JSON.stringify({
                    success: true,
                    network: 'mainnet-beta',
                    programId: PROGRAM_ID.toBase58(),
                    ...data,
                }, null, 2),
            },
        ],
    };
}
function errorResponse(message) {
    return {
        content: [
            {
                type: 'text',
                text: JSON.stringify({ success: false, error: message }),
            },
        ],
    };
}
